File: code/core/backend/engine/executors/javascript1/sdk/lib/plugkv.ts
declare function _pkv_set(tx: number, key: string, value: string, opts?: any): string
declare function _pkv_update(tx: number, key: string, value: string, opts?: any): string
declare function _pkv_get(tx: number, key: string): [PlugValue, string];
declare function _pkv_query(tx: number, opts: PlugQuery): [PlugValue[], string];
declare function _pkv_del(tx: number, key: string): string;
declare function _pkv_batch_del(tx: number, keys: string[]): string;
declare function _pkv_new_txn(): [number, string];
declare function _pkv_rollback(tx: number): string;
declare function _pkv_commit(tx: number): string;


export interface PlugValue {
    key: string
    value: string
    version: number
    // meta
    tag1?: string
    tag2?: string
    tag3?: string
    ttl?: string
}

export interface PlugQuery {
    key_prefix?: string
    load_meta?: boolean
    tag1s?: string[]
    tag2s?: string[]
    tag3s?: string[]
    page_count?: number
    page?: number
}

export interface SetOptions {
    tag1?: string
    tag2?: string
    tag3?: string
    ttl?: number
}

export interface UpdateOptions {
    force_ver?: boolean
    with_version?: boolean
    version?: number
    set_tag1?: boolean
    set_tag2?: boolean
    set_tag3?: boolean
    tag1?: string
    tag2?: string
    tag3?: string
    set_ttl?: boolean
    ttl?: number
}

export class PlugKV {
    txid: number
    constructor(txid: number) {
        this.txid = txid
    }

    quick_get = (key: string): [string, string] => {
        const [val, err] = _pkv_get(this.txid, key)
        if (err) {
            return [null, err]
        }
        return [val.value, null]
    }

    set = (key: string, value: string, opts?: SetOptions): string => {
        return _pkv_set(this.txid, key, value, opts)
    }

    update = (key: string, value: string, opts?: UpdateOptions): string => {
        return _pkv_update(this.txid, key, value, opts)
    }


    get = (key: string): [PlugValue, string] => {
        const [val, err] = _pkv_get(this.txid, key)
        if (err) {
            return [null, err]
        }
        return [val, null]
    }


    query = (opts: PlugQuery): [PlugValue[], string] => {
        return _pkv_query(this.txid, opts)
    }

    delete = (key: string): string => {
        return _pkv_del(this.txid, key)
    }

    batch_delete = (keys: string[]): string => {
        return _pkv_batch_del(this.txid, keys)
    }

    new_txn = (): [PlugKV, string] => {
        const [newtxid, err] = _pkv_new_txn()
        if (err) {
            return [null, err]
        }
        return [new PlugKV(newtxid), null]
    }

    rollback = (): string => {
        return _pkv_rollback(this.txid)
    }

    commit = (): string => {
        return _pkv_commit(this.txid)
    }
}

export const plugkv = new PlugKV(0);

File: code/backend/xtypes/store/plugkv.go
package store

import "github.com/temphia/temphia/code/backend/xtypes/models/entities"

type PlugStateKV interface {
	Set(txid uint32, tenantId, plugId, key, value string, opts *SetOptions) error
	Update(txid uint32, tenantId, plugId, key, value string, opts *UpdateOptions) error
	Get(txid uint32, tenantId, plugId, key string) (*entities.PlugKV, error)
	Query(txid uint32, tenantId, plugId string, query *PkvQuery) ([]*entities.PlugKV, error)
	Del(txid uint32, tenantId, plugId, key string) error
	DelBatch(txid uint32, tenantId, plugId string, keys []string) error

	NewTxn() (uint32, error)
	RollBack(txid uint32) error
	Commit(txid uint32) error
}

type SetOptions struct {
	Tag1 string `json:"tag1,omitempty"`
	Tag2 string `json:"tag2,omitempty"`
	Tag3 string `json:"tag3,omitempty"`
	TTL  int    `json:"ttl,omitempty"`
}

type UpdateOptions struct {
	ForceVer    bool   `json:"force_ver,omitempty"`
	WithVerison bool   `json:"with_version,omitempty"`
	Version     int    `json:"version,omitempty"`
	SetTag1     bool   `json:"set_tag1,omitempty"`
	SetTag2     bool   `json:"set_tag2,omitempty"`
	SetTag3     bool   `json:"set_tag3,omitempty"`
	Tag1        string `json:"tag1,omitempty"`
	Tag2        string `json:"tag2,omitempty"`
	Tag3        string `json:"tag3,omitempty"`
	SetTTL      bool   `json:"set_ttl,omitempty"`
	TTL         int    `json:"ttl,omitempty"`
}

type PkvQuery struct {
	KeyPrefix string   `json:"key_prefix,omitempty"`
	LoadMeta  bool     `json:"load_meta,omitempty"`
	Tag1s     []string `json:"tag1s,omitempty"`
	Tag2s     []string `json:"tag2s,omitempty"`
	Tag3s     []string `json:"tag3s,omitempty"`
	PageCount uint     `json:"page_count,omitempty"`
	Page      uint     `json:"page,omitempty"`
	KeyCursor string   `json:"key_cursor,omitempty"`
}
